oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/ReactiveJWTProcessor.java

```diff
@@ -16,27 +16,10 @@
 
 package org.springframework.security.oauth2.jwt;
 
-import com.nimbusds.jose.jwk.JWK;
-import org.junit.Test;
-
-import java.util.Arrays;
-
-import static org.assertj.core.api.Assertions.*;
-import static org.mockito.Mockito.mock;
-
-/**
- * @author Rob Winch
- * @since 5.1
- */
-public class JWKContextJWKSourceTests {
-	private JWKContextJWKSource source = new JWKContextJWKSource();
-
-	@Test
-	public void getWhenKeysNotEmptyThenContainsKeys() {
-		JWK key = mock(JWK.class);
-		JWKContext jwkContext = new JWKContext(Arrays.asList(key));
-
-		assertThat(this.source.get(null, jwkContext)).containsOnly(key);
-	}
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+import reactor.core.publisher.Mono;
 
+public interface ReactiveJWTProcessor {
```

Comment:
I like all the changes except for this part. I don't like that we have a public interface that has Nimbus types on it. There are a few ideas I had for this:

* Start with making the changes to use the new Nimbus APIs, but do not make a public API available for others. I like this idea a lot (at least for a start since in my mind it is not controversial)
* Move the builder and the interface into `NimbusReactiveJwtDecoder` since it is Nimbus specific. The class/interfaces would include Nimbus in the names since they are Nimbus specific
* Remove the interface and use something more generic like `Converter<Mono<JWTClaimsSet>,SignedJWT>`
* See if we can change the input and output to be something that is not Nimbus specific. For the input we could use a String. However, I'm not sure what we would do for the output. I'd like to avoid writing complex types of our own unless we are going to be using the types in other places.

I think it may be possible to use a combination of these approaches too.

Reply:
I can see the value in the first option, but I figure if we can come to a consensus quickly on the contract, we might as well include it in this PR.

I'd not prefer the second as it would be ideal if the structure were roughly similar to the imperative side of things. We've already seen examples in other places where users have an imperative implementation of something and want to adapt it into their reactive applications. The way I see it, if the builder were part of `NimbusReactiveJwtDecoder`, the output of the builder would be the decoder itself as opposed to a contract similar to Nimbus's `JWTProcessor`, making this kind of reuse trickier.

I like the third option. Just to be sure, though, I think that `Converter<SignedJWT, Mono<JWTClaimSet>>` is correct. This is similar enough to the imperative contract that the user could easily adapt any non-blocking imperative implementation that they already have. It's also a pattern that Spring Security follows in other places, so there is some familiarity with it.

I like the fourth option; however, I tried implementing it and ran into a snag. The `NimbusReactiveJwtDecoder` needs to know the header in order to create an instance of `Jwt`. This means that it needs to parse the String before handing that String off to the processor. When I tried a contract of `<String, Map<String, Object>>` (token -> claim set), I ended up parsing the token twice. Alternatively, I thought I might try `<String, Jwt>` so that the processor can hand back the header as well, but that felt odd since we are now basically repeating the same contract as `NimbusReactiveJwtDecoder`.

Given all of this, I've updated the PR in a separate commit which shows what I see as a working option 3.

Reply:
The second commit you did ends up pretty close to my PR #6367 

However, I got an issue with the `Converter<SignedJWT, Mono<JWTClaimSet>>` 

The code state:

```
@Override
public Mono<Jwt> decode(String token) throws JwtException {
	JWT jwt = parse(token);
	if (jwt instanceof SignedJWT) {
		return this.decode((SignedJWT) jwt);
	}
	throw new JwtException("Unsupported algorithm of " + jwt.getHeader().getAlgorithm());
}
```
Now I personally don't use `EncryptedJWT` (yet????) but it exists. What would be the approach for the people using it ? 

Specialize another `NimbusReactiveJwtDecoder` ?  or extends this one ?  
First my thinking was to create another specialized decoder, but then I though of what can happen on my side:  Multiple authorization server. It means I cannot be 100% sure of what those providers will send me.  This would end up copying/pasting a lot of spring code to make this possible.  

Perhaps it is better to add the `EncryptedJWT` use case, and so what is the better approach  to this ? a constructor with one converter (signed JWT) and another with 2 converter (signed / encrypted) ? a single interface ?

Reply:
I like that you are thinking about these things, but I think we should keep this PR focused on the intended use case. It's better to solve one problem at a time. Spring Security doesn't yet support encrypted JWTs nor Authorization Server multi-tenancy, so trying to add support for it in this PR would probably jump the gun.

However, yes, I believe it would be a new constructor parameter. In the meantime, folks would probably implement `ReactiveJwtDecoder` and write their own reactive encryption support since neither Spring nor Nimbus has yet a reactive JWE API. They can always delegate to Spring Security's reactive JWS support once the JWT is decrypted.